Add rawtr() descriptor for P2TR with specified (tweaked) output key

sipa opened the pull request:
It may be useful to be able to represent P2TR outputs in descriptors whose script tree and/or internal key aren't known. This PR does that, by adding a `rawtr(KEY)` descriptor, where the KEY represents the output key directly. If the private key corresponding to that output key is known, it also permits signing with it.

I'm not convinced this is desirable, but presumably "tr(KEY)" sounds more intended for direct use than "rawtr(KEY)".

sipa committed:
Add rawtr() descriptor for P2TR with unknown tweak

sipa committed:
If P2TR tweaked key is available, sign with it

sipa committed:
Add wallet tests for spending rawtr()

benthecarman commented on doc/descriptors.md:
```diff
@@ -73,6 +73,7 @@ Descriptors consist of several types of expressions. The top level expression is
 - `tr(KEY)` or `tr(KEY,TREE)` (top level only): P2TR output with the specified key as internal key, and optionally a tree of script paths.
 - `addr(ADDR)` (top level only): the script which ADDR expands to.
 - `raw(HEX)` (top level only): the script whose hex encoding is HEX.
+- `rawtr(KEY)` (top level only): P2TR output with the specified key as tweaked key.
```
I think saying `tweaked key` could be confusing unless you know what's going on under the hood of taproot. Maybe something like `output key` would be better

sipa replied:
Agree.

Sjors replied:
Or maybe call it pre-tweaked. I can see how that's more useful than a raw descriptor with the tweaked key.

sipa replied:
Done, renamed to "output key".

sipa commented:
@apoelstra @real-or-random @jonasnick @sanket1729 Perhaps the documentation here needs some disclaimer at least about the potential pitfalls of using it directly to construct a wallet. It's not intended for that purpose, only for things like representing otherwise incomplete knowledge in e.g. inference, but adding it to the descriptor language does make it available for all purposes. Can you help think about examples/pitfalls that could be listed?

sanket1729 commented:
Agreed that the documentation needs a disclaimer. The two drawbacks that I can think of.

1) This one is listed in the BIP 341: If people directly use output keys in aggregation using proofs of knowledge.(https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_ref-22-0)
2) In future protocol constructions, it might be desirable to prove that there is no script path given TxOut/ScriptPubKey/Address. This is not possible when using rawtr(), but is possible with tr() by revealing the internal key.

Some more points for discussion:

- Can we separate descriptors into two types in the spec itself? So, we can have `addr`, `raw`, `rawtr` into the infer-only category. The rest into another category that we can sign. In the infer only category, we can restrict the KEY expression to be a concrete public key without any BIP32 derivations or wildcards.

- Correct me if I am wrong if the only use of inferred descriptors is that I can give it another party for watching a particular txout for me without revealing/knowing how the txout was constructed. So, assuming we disallow signing `rawtr()` (to encourage the usage of `tr` ) is there is any use of `rawtr()` over `raw()`?

real-or-random commented:
The only pitfall I see is the one mentioned by @sanket1729 : `rawtr()` is potentially incomplete information because there could always be another way to spend that you don't know about... But that's exactly its feature, so I guess it's fine. Anyway, I think if you do "raw" stuff, you should be know what you're doing.

Sjors commented:
> * is there is any use of `rawtr()` over `raw()`?

I second that question.

sipa commented:
Right; I guess the only advantage of rawtr over raw/addr is dealing with cases where the public key is either actually untweaked (and e.g. you need want to provide origin information along with it) or it is tweaked in an unknown way, but you have the tweaked private key.

raw() in those cases cannot represent all information you have, while rawtr can.

apoelstra commented:
Agree with Sanket's comments. I also agree that having `raw` in the name should be sufficient warning for users to avoid this, especially since it's not any harder to use `tr`.

Regarding the usefulness, I think it's definitely useful for "infer-only" cases like running `scanutxoset`. May also be useful for recovery scenarios ... although it's hard for me to think of a case where you've generated a weird Taproot key but somehow need Bitcoin Core to spend it.

sipa commented:
@apoelstra https://twitter.com/achow101/status/1459617340123926534 may or may not have been related. Arguably, not a good justification, but a justification nonetheless...

apoelstra commented:
Not the worst usecase :) I use vanity addresses in Liquid test networks sometimes to make logs easier to read and having descriptor support in Core could plausibly make that better.

sanket1729 commented:
A [usecase](https://gist.github.com/Kixunil/0ddb3a9cdec33342b97431e438252c0a) for `rawtr`. Note that this protocol is not reviewed(and might be insecure), but I am starting to think we don't want to preclude people from using new ways to create output keys. In this protocol, it is still possible to prove that there is no script spend using a different way than revealing the internal key. Such proof is not required for the correctness above protocol, but it highlights that there might be other ways to prove the non-existence of script spend than revealing the internal key. 

Although as I mentioned later in response [comment]( https://gist.github.com/Kixunil/0ddb3a9cdec33342b97431e438252c0a#gistcomment-4013170), this can also be made compatible with `tr` descriptor. But doing so would be inefficient and it offers no benefit because we can already prove that there is no script spend. 

Overall, I am leaning towards a concept ACK for this.

w0xlt reviewed (commented):
Should `tr(xpriv)` and `rawtr(xpriv)` return different addresses ?

I created two blank wallets and imported `tr(xpriv)` into one and `rawtr(xpriv)` into the other. The `xpriv` was the same.

The addresses have the same `desc` property but not the same `address`, as seen below.

There are also some fields missing from the address generated from `rawtr(xpriv)` like `timestamp`.

```bash
$ ./src/bitcoin-cli -regtest -named createwallet wallet_name=tr_test blank=true

$ ./src/bitcoin-cli -regtest importdescriptors "[{\"desc\": \"tr(tprv8ZgxMBicQKsPdKziibn63Rm6aNzp7dSjDnufZMStXr71Huz7iihCRpbZZZ6Voy5HyuHCWx6foHMipzMzUq4tZrtkZ24DJwz5EeNWdsuwX5h/*)#223d6gp6\", \"active\": true, \"timestamp\": \"now\"}]"

$ ./src/bitcoin-cli -regtest -named createwallet wallet_name=rawtr_test blank=true

$ ./src/bitcoin-cli -regtest -rpcwallet="rawtr_test" importdescriptors "[{\"desc\": \"rawtr(tprv8ZgxMBicQKsPdKziibn63Rm6aNzp7dSjDnufZMStXr71Huz7iihCRpbZZZ6Voy5HyuHCWx6foHMipzMzUq4tZrtkZ24DJwz5EeNWdsuwX5h/*)#223d6gp6\", \"active\": true, \"timestamp\": \"now\"}]"

$ ./src/bitcoin-cli -regtest -rpcwallet="tr_test" -named getnewaddress address_type='bech32m'
bcrt1p8eafa4hnyfae7kyc42ehndduu49zfuhneerma58j2xexc6lx5wxsm8e5j7

$ ./src/bitcoin-cli -regtest -rpcwallet="rawtr_test" -named getnewaddress address_type='bech32m'
bcrt1pqwnxn65jduuptqhvfgqqh9rjh29pwdrlt7c4nmwafgrsx6n8rr4supy9dh

$ ./src/bitcoin-cli -regtest -rpcwallet="rawtr_test" getaddressinfo "bcrt1pqwnxn65jduuptqhvfgqqh9rjh29pwdrlt7c4nmwafgrsx6n8rr4supy9dh"
{
  "address": "bcrt1pqwnxn65jduuptqhvfgqqh9rjh29pwdrlt7c4nmwafgrsx6n8rr4supy9dh",
  "scriptPubKey": "512003a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb",
  "ismine": true,
  "solvable": true,
  "desc": "rawtr([8324d0e9/0]03a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb)#5rk6yanl",
  "parent_desc": "rawtr(tpubD6NzVbkrYhZ4Wo2WcFSgSqRD9QWkGxddo6WSqsVBx7uQ8QEtM7WncKDRjhFEexK119NigyCsFygA4b7sAPQxqebyFGAZ9XVV1BtcgNzbCRR/*)#x2xz0lq0",
  "iswatchonly": false,
  "isscript": true,
  "iswitness": true,
  "witness_version": 1,
  "witness_program": "03a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb",
  "ischange": false,
  "labels": [
    ""
  ]
}

$ ./src/bitcoin-cli -regtest -rpcwallet="tr_test" getaddressinfo "bcrt1p8eafa4hnyfae7kyc42ehndduu49zfuhneerma58j2xexc6lx5wxsm8e5j7"
{
  "address": "bcrt1p8eafa4hnyfae7kyc42ehndduu49zfuhneerma58j2xexc6lx5wxsm8e5j7",
  "scriptPubKey": "51203e7a9ed6f3227b9f5898aab379b5bce54a24f2f3ce47bed0f251b26c6be6a38d",
  "ismine": true,
  "solvable": true,
  "desc": "tr([8324d0e9/0]03a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb)#05zmkxzl",
  "parent_desc": "tr(tpubD6NzVbkrYhZ4Wo2WcFSgSqRD9QWkGxddo6WSqsVBx7uQ8QEtM7WncKDRjhFEexK119NigyCsFygA4b7sAPQxqebyFGAZ9XVV1BtcgNzbCRR/*)#fdlr9alf",
  "iswatchonly": false,
  "isscript": true,
  "iswitness": true,
  "witness_version": 1,
  "witness_program": "3e7a9ed6f3227b9f5898aab379b5bce54a24f2f3ce47bed0f251b26c6be6a38d",
  "ischange": false,
  "timestamp": 1646090838,
  "hdkeypath": "m/0",
  "hdseedid": "0000000000000000000000000000000000000000",
  "hdmasterfingerprint": "8324d0e9",
  "labels": [
    ""
  ]
}

sipa commented:
@w0xlt Yes, that'd definitely expected. `rawtr(KEY)` literally puts the key in the scriptPubKey directly. `tr(KEY)` will construct an output key derived from KEY as internal key.

If the two weren't different, we wouldn't need to add `rawtr` in the first place.

w0xlt commented:
@sipa Thanks for the clarification.

But since the wallet is able to generate multiple addresses for `rawtr(KEY)`, I assumed that there would also be key derivations for KEY in this case.

For example:
```bash
# First address
"desc":    "tr([8324d0e9/0]03a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb)#05zmkxzl"
"desc": "rawtr([8324d0e9/0]03a669ea926f381582ec4a000b9472ba8a17347f5fb159eddd4a07036a6718eb)#5rk6yanl"

# Second address
"desc":    "tr([8324d0e9/1]bbf56b14b119bccafb686adec2e3d2a6b51b1626213590c3afa815d1fd36f85d)#0djkz4yx"
"desc": "rawtr([8324d0e9/1]bbf56b14b119bccafb686adec2e3d2a6b51b1626213590c3afa815d1fd36f85d)#56xhsw4x"
```

Shouldn't `rawtr()` generate just a single address ?

sipa commented:
Different keys give rise to different addresses....

sipa commented:
Rebased.

w0xlt commented:
It looks like the CI is failing due to the difference between the expected result and the RPC result in `rpc_decodescript.py:decodescript_datadriven_tests()`.

Script (first element in `rpc_decodescript.json`):
`5120eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee`

Expected result:
`{'asm': '1 eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee', 'address': 'bcrt1pamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhqz6nvlh', 'desc': 'addr(bcrt1pamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhqz6nvlh)#v52jnujz', 'type': 'witness_v1_taproot'}`

RPC result:
`{'asm': '1 eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee', 'desc': 'rawtr(eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)#jk7c6kys', 'address': 'bcrt1pamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhwamhqz6nvlh', 'type': 'witness_v1_taproot'}`

w0xlt commented:
`wallet_taroot.py` is also failing. Apparently the PSBT is not completing and `assert(res['complete'])` fails.

sipa commented:
I'll get back to this soon.

w0xlt commented:
I noticed the behavior below. I'm not sure if it's as expected.

The address `bcrt1...8n6wyskueguk` returns the  `tr(...)` descriptor in `getaddressinfo` RPC, which looks correct as it was generated via `getnewaddress '' 'bech32m'`
But when this same address is used as the recipient of a transaction, its descriptor is displayed as `rawtr(...)`, as seen below.

```bash
$ bitcoin-cli -regtest -rpcwallet="receiver" getnewaddress '' 'bech32m'
bcrt1pwlhw9uxkga774g6er3hthkaa4npvr2zcutt4gcqxyf5f2y8n6wyskueguk

$ bitcoin-cli -regtest -rpcwallet="receiver" getaddressinfo "bcrt1pwlhw9uxkga774g6er3hthkaa4npvr2zcutt4gcqxyf5f2y8n6wyskueguk"
{
  "address": "bcrt1pwlhw9uxkga774g6er3hthkaa4npvr2zcutt4gcqxyf5f2y8n6wyskueguk",
  "scriptPubKey": "512077eee2f0d6477deaa3591c6ebbdbbdacc2c1a858e2d754600622689510f3d389",
  ...
  "desc": "tr([92c4c112/86'/1'/0'/0/0]a612d64e3415afb653e98da12bb34a9c0785a10c6880cbf0436791a89428826b)#f67tyh68",
  ...
}

$ bitcoin-cli -regtest -rpcwallet="sender" -named send outputs="{\"bcrt1pwlhw9uxkga774g6er3hthkaa4npvr2zcutt4gcqxyf5f2y8n6wyskueguk\": 12.00000381}" fee_rate=1 options="{\"add_to_wallet\": false }"
{
  ...
  "hex": "020000...ca3078fbd0900000000"
}

$ bitcoin-cli -regtest decodetransaction "020000...ca3078fbd0900000000"
{
  "txid": "fa4f78fc970bbae4e932690b5b9370d18545316f43fa126991a0e76cf57a1e63",
  "hash": "dc5ac5d41fd0cf0b4990f9cafb3b64eac5b05680776e80c61f3867ee51b70df2",
  "version": 2,
  ...
  "vout": [
    {
      "value": 12.00000381,
      "n": 0,
      "scriptPubKey": {
        "asm": "1 77eee2f0d6477deaa3591c6ebbdbbdacc2c1a858e2d754600622689510f3d389",
        "desc": "rawtr(77eee2f0d6477deaa3591c6ebbdbbdacc2c1a858e2d754600622689510f3d389)#u26fgmea",
        "hex": "512077eee2f0d6477deaa3591c6ebbdbbdacc2c1a858e2d754600622689510f3d389",
        "address": "bcrt1pwlhw9uxkga774g6er3hthkaa4npvr2zcutt4gcqxyf5f2y8n6wyskueguk",
        "type": "witness_v1_taproot"
      }
    },
    ...
  ]
}
```

sipa commented:
@w0xlt Yes, that's expected. The sender doesn't know the internal public key the receiver wants, so it can only show the tweaked key.

theStack commented:
Concept ACK

sipa commented:
Rebased.

sipa commented:
Rebased.

real-or-random commented:
The tests fails, see https://cirrus-ci.com/task/4834974200233984?logs=ci#L3424

sipa commented:
Rebased, and hopefully fixed test failures.

sanket1729 commented on src/script/descriptor.cpp:
```diff
@@ -917,6 +917,24 @@ class TRDescriptor final : public DescriptorImpl
+        return Vector(GetScriptForDestination(output));
+    }
+public:
+    RawTRDescriptor(std::unique_ptr<PubkeyProvider> internal_key) : DescriptorImpl(Vector(std::move(internal_key)), "rawtr") {}
```
nit: here and elsewhere

Perhaps better renamed as `output_key` over `internal_key`

sipa replied:
Done.

sanket1729 reviewed (approved):
ACK left one small nit. Feel free to ignore

w0xlt reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/23480/commits/9342970d7253abbf97c2bde6c242e29d9445c311.

This `rawtr` implementation is used in PR https://github.com/bitcoin/bitcoin/pull/24897 and has been successfully stored and spent on signet.

real-or-random commented:
> Perhaps the documentation here needs some disclaimer at least about the potential pitfalls of using it directly to construct a wallet.

I think the doc part is still unresolved?

w0xlt commented:
This might fix the CI errors:
In test/functional/wallet_taproot.py (https://github.com/bitcoin/bitcoin/pull/23480/commits/9342970d7253abbf97c2bde6c242e29d9445c311):

```diff
  from test_framework.key import H_POINT
+ from test_framework.segwit_addr import encode_segwit_address
```

sipa commented:
Rebased, and made a few changes:
* Added a warning to doc/descriptors.md, and some other changes, following https://github.com/bitcoin/bitcoin/pull/23480#issuecomment-1159573267 by @real-or-random.
* Fixed test failure, following https://github.com/bitcoin/bitcoin/pull/23480#issuecomment-1186014033 by @w0xlt.
* Renamed internal_key -> output_key following https://github.com/bitcoin/bitcoin/pull/23480#discussion_r900514564 by @sanket1729.

sanket1729 reviewed (approved):
code review ACK 544b4332f0e122167bdb94dc963405422faa30cb

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/23480/commits/544b4332f0e122167bdb94dc963405422faa30cb

achow101 commented:
ACK 544b4332f0e122167bdb94dc963405422faa30cb
